File: CHUONG_5/5.10_RecyclerView_Java/app/src/main/java/com/example/myapplication/ui/MainActivity.java
package com.example.myapplication.ui;

import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.RecyclerView;

import com.example.myapplication.R;
import com.example.myapplication.data.source.StudentDataRepository;
import com.example.myapplication.data.source.local.LocalDataSource;
import com.example.myapplication.ui.adapter.StudentAdapter;
import com.example.myapplication.ui.viewmodel.StudentViewModel;
import com.example.myapplication.ui.viewmodel.StudentViewModelFactory;

import java.util.Objects;

public class MainActivity extends AppCompatActivity {
    private StudentAdapter mAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        setupViews();
        setupViewModel();
    }

    private void setupViews() {
        RecyclerView recyclerView = findViewById(R.id.recycler_student);
        StudentAdapter.OnItemClickListener callback = student -> {
            String message = getString(R.string.message_info, student.getFullName());
            Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
        };
        mAdapter = new StudentAdapter(null, callback);
        recyclerView.setAdapter(mAdapter);
        DividerItemDecoration divider =
                new DividerItemDecoration(this, DividerItemDecoration.VERTICAL);
        Drawable drawableDivider = Objects.requireNonNull(
                AppCompatResources.getDrawable(this, R.drawable.list_divider));
        divider.setDrawable(drawableDivider);
        recyclerView.addItemDecoration(divider);
    }

    private void setupViewModel() {
        LocalDataSource localDataSource = new LocalDataSource(this);
        StudentDataRepository repository = new StudentDataRepository(localDataSource);
        StudentViewModel viewModel = new ViewModelProvider(
                this, new StudentViewModelFactory(repository)).get(StudentViewModel.class);
        viewModel.student.observe(this, mAdapter::updateData);
    }
}

File: CHUONG_5/5.11_RecyclerView_Kotlin/app/src/main/java/com/example/myapplication/MainActivity.kt
package com.example.myapplication

import android.os.Bundle
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.RecyclerView
import com.example.myapplication.LocalDataSource


class MainActivity : AppCompatActivity() {
    private lateinit var recyclerStudent: RecyclerView
    private lateinit var studentAdapter: StudentAdapter
    private lateinit var viewModel: StudentViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }
        setupView()
        setupViewModel()
    }

    private fun setupView() {
        recyclerStudent = findViewById(R.id.rv_student)
        studentAdapter = StudentAdapter(object : StudentAdapter.OnStudentClickListener {
            override fun onClick(student: Student) {
                Toast.makeText(
                    this@MainActivity,
                    student.fullName.toString(),
                    Toast.LENGTH_SHORT
                )
                    .show()
            }
        })
        recyclerStudent.adapter = studentAdapter
        val divider = DividerItemDecoration(this, DividerItemDecoration.VERTICAL)
        recyclerStudent.addItemDecoration(divider)
    }

    private fun setupViewModel() {
        val dataSource = LocalDataSource(this)
        val viewModelFactory = StudentViewModel.Factory(dataSource)
        viewModel = ViewModelProvider(this, viewModelFactory)[StudentViewModel::class.java]
        viewModel.loadStudents()
        viewModel.students.observe(this) {
            studentAdapter.updateStudents(it)
        }
    }
}